canal/sync.go

```diff
@@ -122,42 +114,38 @@ func (c *Canal) runSyncBinlog() error {
 			if e.GSet != nil {
 				c.master.UpdateGTIDSet(e.GSet)
 			}
-			var (
-				mb    [][]byte
-				db    []byte
-				table []byte
-			)
-			regexps := []regexp.Regexp{*expCreateTable, *expAlterTable, *expRenameTable, *expDropTable, *expTruncateTable}
-			for _, reg := range regexps {
-				mb = reg.FindSubmatch(e.Query)
-				if len(mb) != 0 {
-					break
-				}
-			}
-			mbLen := len(mb)
-			if mbLen == 0 {
+			stmts, _, err := c.parser.Parse(string(e.Query), "", "")
+			if err != nil {
+				log.Errorf("parse query err %v", err)
 				continue
 			}
-
-			// the first last is table name, the second last is database name(if exists)
-			if len(mb[mbLen-2]) == 0 {
-				db = e.Schema
-			} else {
-				db = mb[mbLen-2]
-			}
-			table = mb[mbLen-1]
-
-			savePos = true
-			force = true
-			c.ClearTableCache(db, table)
-			log.Infof("table structure changed, clear table cache: %s.%s\n", db, table)
-			if err = c.eventHandler.OnTableChanged(string(db), string(table)); err != nil && errors.Cause(err) != schema.ErrTableNotExist {
-				return errors.Trace(err)
-			}
-
-			// Now we only handle Table Changed DDL, maybe we will support more later.
-			if err = c.eventHandler.OnDDL(pos, e); err != nil {
-				return errors.Trace(err)
+			for _, stmt := range stmts {
+				switch t := stmt.(type) {
+				case *ast.RenameTableStmt:
+					for _, tableInfo := range t.TableToTables {
+						db := tableInfo.OldTable.Schema.String()
+						table := tableInfo.OldTable.Name.String()
+						c.updateTable(db, table, e, pos, ev.Header.Timestamp)
+					}
+				case *ast.AlterTableStmt:
+					db := t.Table.Schema.String()
```

Comment:
need to handle `alter table xxx rename to yyy`

Reply:
this case it is just same as alter table. AlterTableStmt also do not have a member of `TableToTables `.
in this case, we just need  to clear the older table cache, this is what we do  already.